src/sagemaker_xgboost_container/data_utils.py

```diff
@@ -528,33 +528,50 @@ def _get_pipe_mode_files_path(data_path: Union[List[str], str]) -> List[str]:
     return files_path
 
 
+def _make_symlinks_for_files_under_a_folder(dest_path: str, data_path: str):
+    if (not os.path.exists(dest_path)) or (not os.path.exists(data_path)):
+        raise exc.AlgorithmError("Unable to create symlinks as {data_path} or {dest_path} doesn't exist ")
+
+    logging.info("Making smlinks from folder {} to folder {}".format(data_path, dest_path))
+
+    if os.path.isfile(data_path):
+        _make_symlink(data_path, dest_path, os.path.basename(data_path))
+        return
+
+    else:
+        for item in os.scandir(data_path):
+            if item.is_file():
+                _make_symlink(item.path, dest_path, item.name)
+            elif item.is_dir():
+                _make_symlinks_for_files_under_a_folder(dest_path, item.path)
+
+
 def _get_file_mode_files_path(data_path: Union[List[str], str]) -> List[str]:
     """
     :param data_path: Either directory or file
     """
     # In file mode, we create a temp directory with symlink to all input files or
     # directories to meet XGB's assumption that all files are in the same directory.
 
+    logging.info("File path {} of input files".format(data_path))
+    # Create a directory with symlinks to input files.
+    files_path = "/tmp/sagemaker_xgboost_input_data"
+    shutil.rmtree(files_path, ignore_errors=True)
+    os.mkdir(files_path)
     if isinstance(data_path, list):
-        logging.info("File path {} of input files".format(data_path))
-        # Create a directory with symlinks to input files.
-        files_path = "/tmp/sagemaker_xgboost_input_data"
-        shutil.rmtree(files_path, ignore_errors=True)
-        os.mkdir(files_path)
-        for index, path in enumerate(data_path):
-            if not os.path.exists(path):
-                return None
-            if os.path.isfile(path):
-                _make_symlink(path, files_path, os.path.basename(path), index)
-            else:
-                for file in os.scandir(path):
-                    _make_symlink(file, files_path, file.name, index)
-
+        for path in data_path:
+            _make_symlinks_for_files_under_a_folder(files_path, path)
     else:
         if not os.path.exists(data_path):
             logging.info("File path {} does not exist!".format(data_path))
             return None
-        files_path = get_files_path_from_string(data_path)
+        elif os.path.isdir(data_path):
+            # traverse all sub-dirs to load all training data
+            _make_symlinks_for_files_under_a_folder(files_path, data_path)
+        elif os.path.isfile(data_path):
+            files_path = data_path
```

Comment:
Can we just call `_make_symlinks_for_files_under_a_folder ` here? Isn't it able to handle singular files:?

Reply:
yes, it 's able to do that..Let me refactor the info message in next commit to prevent confusion.